react/index.d.ts

```diff
@@ -165,8 +165,8 @@ declare namespace React {
     class Component<P, S> implements ComponentLifecycle<P, S> {
         constructor(...args: any[]);
         constructor(props?: P, context?: any);
-        setState(f: (prevState: S, props: P) => S, callback?: () => any): void;
-        setState(state: S, callback?: () => any): void;
+        setState<K extends keyof S>(f: (prevState: S, props: P) => Pick<S, K> | S, callback?: () => any): void;
+        setState<K extends keyof S>(state: Pick<S, K> | S, callback?: () => any): void;
```

Comment:
Yes that is correct. I'll put it in the README and the definition file. Good call.